tests: Enable attestation e2e tests for IBM SE

BbolroC opened the pull request:
This PR enables [k8s-confidential-attestation.bats](https://github.com/kata-containers/kata-containers/blob/main/tests/integration/kubernetes/k8s-confidential-attestation.bats) for s390x. Detailed explanation can be found in each commit message.

Fixes: #9933

For reviewers: the tests have been passing since the 27th of June at https://github.com/kata-containers/kata-containers/actions/workflows/ci-nightly-s390x.yaml

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

stevenhorsman commented on tools/packaging/static-build/coco-guest-components/build.sh:
Based on the logic in the guest-components `Makefile`, doesn't setting `TEE_PLATFORM` effectively overide what is set in `ATTESTER`?

BbolroC replied:
yeah, setting `TEE_PLATFORM` effectively overrides `ATTESTER`.

stevenhorsman commented on tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh:
```diff
@@ -114,6 +114,7 @@ docker run \
 	--env ARTEFACT_REGISTRY_PASSWORD="${ARTEFACT_REGISTRY_PASSWORD}" \
 	--env TARGET_BRANCH="${TARGET_BRANCH}" \
 	--env BUILDER_REGISTRY="${BUILDER_REGISTRY}" \
+	--env TEE_PLATFORM="${TEE_PLATFORM:-}" \
```
Shouldn't we extract out `TEE_PLATFORM="${TEE_PLATFORM:-}"` like all the other envs?

BbolroC replied:
If we don't have this line, there is no way of building the guest components where `se-attester` is solely built with `make coco-guest-components-tarball`. With this line, we can achieve to have those guest components by running `make coco-guest-components-tarball TEE_PLATFORM=se`. 

If there is another way of achieving this goal, please let me know. I will following that way.

stevenhorsman replied:
Sorry for not being clear. The pattern in that file seems to be to set `TEE_PLATFORM="${TEE_PLATFORM:-}"` separately (e.g. https://github.com/kata-containers/kata-containers/blob/2396968d92f00636caa216675a5b2e0f85844742/tools/packaging/kata-deploy/local-build/kata-deploy-binaries-in-docker.sh#L82-L103), then in the `docker` call, just use `--env TEE_PLATFORM="${TEE_PLATFORM}"`. I don't think the effect will be different and don't necessarily even like that convention, but I think it's best to be consistent.

BbolroC replied:
yes, you are right. It does not look consistent with others. Thanks!

stevenhorsman commented on tools/packaging/static-build/coco-guest-components/build.sh:
```diff
@@ -39,17 +39,17 @@ docker pull ${container_image} || \
 # Temp settings until we have a matching TEE_PLATFORM
-TEE_PLATFORM=""
-RESOURCE_PROVIDER="kbs,sev"
+TEE_PLATFORM="${TEE_PLATFORM:-all}"
```
If we default to `all` here, are we not going to end up trying to build the tpm attesters in `s390x` and fail here, or is there an override that sets `TEE_PLATFORM=se` if the arch it s390x, that I'm missing?

BbolroC replied:
You are right. That should make the s390x build failing. There is no override setting `TEE_PLATFORM=se` in the s390x pipeline. It must be `TEE_PLATFORM="${TEE_PLATFORM:-}"`.  Thanks for pointing it out.

stevenhorsman reviewed (approved):
The code looks good to me now, lets hope the CI is okay 🤞. Thanks for the updates!

wainersm commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -237,7 +237,13 @@ function kbs_k8s_deploy() {
 	# Tests should fill kbs resources later, however, the deployment
 	# expects at least one secret served at install time.
-	echo "somesecret" > overlays/key.bin
+	if [ -d "overlays/$(uname -m)" ]; then
```
Hi @BbolroC !

Kata uses trustee v0.9.0 which doesn't contain the multi-arch support. So once we bump to newest trustee we will re-do this change? i.e. to always `echo "somesecret" > overlays/$(uname -m)/key.bin` ? 

If yes, could we mark with a `TODO` comment?

Thinking aloud... perhaps trustee's `deploy-kbs.sh` could deal with that secret. Consumers should simply export the secret in a variable, this way any break on trustee side doesn't affect us.

wainersm replied:
ah, also any special reason to not bump trustee on this PR?

stevenhorsman replied:
I guess that the trustee team cut the `v0.9.0` as they felt it was stable. That doesn't mean we can't bump it here, but we probably want to get some input/agreement with them first.

BbolroC replied:
Thanks for the comment. I think the change covers both cases (a version with/without the multi-arch support), which means x86_64 is not affected as the version changes. However, we could refactor the code if `echo "somesecret" > overlays/key.bin` is no longer used. 

> Thinking aloud... perhaps trustee's deploy-kbs.sh could deal with that secret. Consumers should simply export the secret in a variable, this way any break on trustee side doesn't affect us.

I was thinking the same. But even in this case, we will face the similar issue once when bumping the version between two repos.

ATM, the s390x nightly test is using the main branch for trustee like:

```diff
diff --git a/tests/integration/kubernetes/confidential_kbs.sh b/tests/integration/kubernetes/confidential_kbs.sh
index 36f4244b5..1e297eaf1 100644
--- a/tests/integration/kubernetes/confidential_kbs.sh
+++ b/tests/integration/kubernetes/confidential_kbs.sh
@@ -230,10 +230,10 @@ function kbs_k8s_deploy() {

        echo "::group::Clone the kbs sources"
        git clone --depth 1 "${repo}" "$COCO_TRUSTEE_DIR"
-       pushd "$COCO_TRUSTEE_DIR"
-       git fetch --depth=1 origin "${version}"
-       git checkout FETCH_HEAD -b kbs_$$
-       popd
+       #pushd "$COCO_TRUSTEE_DIR"
+       #git fetch --depth=1 origin "${version}"
+       #git checkout FETCH_HEAD -b kbs_$$
+       #popd
        echo "::endgroup::"

        pushd "${COCO_KBS_DIR}/config/kubernetes/"
```

I've opened a thread for the update of the trustee for kata containers on Slack: https://cloud-native.slack.com/archives/C039JSH0807/p1720080885154259. If the trustee team is okay to bump it, I will include it here. Thanks.

wainersm commented on tools/packaging/guest-image/build_se_image.sh:
```diff
@@ -9,27 +9,18 @@ set -o errexit
-readonly script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-readonly packaging_root_dir="$(cd "${script_dir}/../" && pwd)"
-readonly kata_root_dir="$(cd "${packaging_root_dir}/../../" && pwd)"
+script_name="$(basename "${BASH_SOURCE[0]}")"
```
I'm curious why it needed the removal of `readonly`

wainersm replied:
hmm maybe because same variables are set in [tools/packaging/guest-image/lib_se.sh](https://github.com/kata-containers/kata-containers/pull/9951/commits/15b16041850acd7cd744a5a758c5212048820e30#diff-3513f55f212771374656907cdec9a7f2335c41d9903426eb6687dacc307f1510) so it complains they are readonly?

BbolroC replied:
Exactly! 😉 They are declared as readonly before sourcing the library. Fortunately, the two files are located in the same directory. We could differentiate those names to prevent this kind of conflict if the library is used in various places, which doesn't seem to be the case here. I need your advise on this. Thanks!

wainersm commented on tools/packaging/guest-image/lib_se.sh:
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
```
didn't review this file. I assume it's just a copy without modification ;)

BbolroC replied:
The only change I made was to relocate `rm -f "${parmfile}"` from a trap callback to the end of the `build_secure_image()` function. Thanks!

BbolroC commented:
@wainersm  As confirmed by the trustee team, I've added a commit to bump trustee and updated the commit for changing the location where a `key.bin` is created. Please take a look. Thanks!

BbolroC commented:
The following issue have been observed:

```
Setting up custom PCCS for TDX
  error: accumulating resources: accumulation err='accumulating resources from '../nodeport': '/tmp/trustee/kbs/config/kubernetes/nodeport' must resolve to a file': couldn't make target for path '/tmp/trustee/kbs/config/kubernetes/nodeport': unable to find one of 'kustomization.yaml', 'kustomization.yml' or 'Kustomization' in directory '/tmp/trustee/kbs/config/kubernetes/nodeport'
```

This should be resolved from trustee. This PR should be held until then. Thanks.

fitzthum commented:
Ok, looks like there is a little bug in the most recent version of Trustee. We will break TDX if we bump it. I should be able to post a fix soon tho.

fitzthum commented:
See https://github.com/confidential-containers/trustee/pull/439

wainersm commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -310,6 +321,29 @@ EOF
+				return 1
+			fi
+			creds_dir="${IBM_SE_CREDS_DIR}"
+			kernel_params_value="agent.guest_components_rest_api=resource agent.aa_kbc_params=cc_kbc::${svc_host}"
```
Hi @BbolroC !

I started reading this commit with a feeling of "it seems to belong to somewhere else". Then I hit that line, which made me more inclined towards suggesting to wrap in a function to be called from the test. Why? Because if this `kernel_params_value` should match the one it's passed via Pod annotation, then it might be different from test to test.

For example, the recent added [test for sealed secrets](https://github.com/kata-containers/kata-containers/blob/main/tests/integration/kubernetes/k8s-sealed-secret.bats#L36) will pass the following kernel parameters instead: `agent.guest_components_procs=confidential-data-hub  agent.aa_kbc_params=cc_kbc::${svc_host}`

or am I missing something?

BbolroC replied:
Yes, from the beginning, this was wrapped as a function and called from k8s-confidential-attestation.bats. But, at the time of refactoring, I realised that file is the only one calling the function. That is why I had this snippet called once when the KBS is deployed. 

I totally agree with your opinion. I will change the code.

wainersm commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -310,6 +321,29 @@ EOF
+				>&2 echo "ERROR: config file not found: ${config_file_path}"
+				return 1
+			fi
+			kata_base_dir=$(dirname $(kata-runtime --config ${config_file_path} env --json | jq -r '.Kernel.Path'))
```
`kernel_base_dir`?

I was going to suggest you to use `extract_kata_env()` from `tests/common.bash`, but it would need to add support to export the kernel path. I don't know if that function will get the right config file path too. It's up to you fight that battle or not :)

BbolroC replied:
Thanks for the suggestion. I think renaming it to `kernel_base_dir` would be enough in this case. Thanks.

wainersm commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -456,3 +490,30 @@ _handle_ingress_nodeport() {
+		>&2 echo "ERROR: HKD_PATH is not set"
+		return 1
+	fi
+	ibmse_creds_dir=$(mktemp -d -t ibmse.creds.XXXXXXXXXX)
```
if you end up moving this code close to the tests then you will be able to use bats' temporary directory and not care about cleaning that directory later.

BbolroC replied:
A lifecycle of this directory should match with that of KBS rather than the tests. It is mounted to the KBS as a persistent volume and cleaned up by the post-action script. If it is removed at a task `running tests`, this would be problematic for uninstalling a cluster. I would like to leave it as it is. Thanks.

BbolroC replied:
In line with my argument, I took out the declaration of `IBM_SE_CREDS_DIR`. The env variable should be defined/cleaned up in the workflow for the internal SE k8s tests.

BbolroC replied:
I made a new function `repack_secure_image()` general to cover a case where a secure image only needs to be built without KBS. Thanks!

wainersm commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -456,3 +490,30 @@ _handle_ingress_nodeport() {
+	export IBM_SE_CREDS_DIR="${ibmse_creds_dir}"
+	pushd "${ibmse_creds_dir}"
+	mkdir {certs,crls,hdr,hkds,rsa}
+	openssl genrsa -aes256 -passout pass:test1234 -out encrypt_key-psw.pem 4096
```
I hate that many websites don't allow me to set such as amazing test1234 password!

wainersm commented:
Hi @BbolroC ! I think I finished my review. Left just a few comments, overall looks great!

BbolroC commented:
I've confirmed the e2e tests for the latest update are passed.

fitzthum commented:
TDX and SEV/SNP CI failing to rebase :'(

GabyCT commented:
I just re-run the TDX CI job

BbolroC commented:
> I just re-run the TDX CI job

The job has not been picked up by the runner quite a while. Is the runner being occupied too much?

GabyCT commented:
> > I just re-run the TDX CI job
> 
> The job has not been picked up by the runner quite a while. Is the runner being occupied too much?

yes, the job is in queue I re-triggered other PRs for TDX and seems like now is running another PR

BbolroC commented:
@GabyCT The error says:

```
The pkg-config command could not be found.

  Most likely, you need to install a pkg-config package for your OS.
  Try `apt install pkg-config`, or `yum install pkg-config`,
  or `pkg install pkg-config`, or `apk add pkgconfig` depending on your distribution.
```

The package seems not to be installed. Or am I missing something?

GabyCT commented:
> @GabyCT The error says:
> 
> ```
> The pkg-config command could not be found.
> 
>   Most likely, you need to install a pkg-config package for your OS.
>   Try `apt install pkg-config`, or `yum install pkg-config`,
>   or `pkg install pkg-config`, or `apk add pkgconfig` depending on your distribution.
> ```
> 
> The package seems not to be installed. Or am I missing something?

@BbolroC I just installed the package and re-triggered the job

GabyCT commented:
@BbolroC I installed the required packages to avoid the failures in installing KBS client step, however, I re-ran the job, and I got twice the same failure on TDX on the following test
```
[run_kubernetes_tests.sh:122] ERROR: Tests FAILED from suites: k8s-confidential.bats
```
```
#   Warning  Failed     82s               kubelet            Error: failed to create containerd container: error unpacking image: failed to prepare extraction snapshot "extract-304855051-hXUq sha256:414cb028a051f35dcb03e7e474dabfa2d04492a94e30b4862d819cab4917cc01": create instance 6186: object with key "6186" already exists: unknown
```

fitzthum commented:
> seems like this failure is not related with this PR. @wainersm @fitzthum @stevenhorsman if you are familiar with this kind of failure?

pretty common snapshotter gremlin. @ChengyuZhu6 probably knows how to reset things to fix this

BbolroC commented:
@wainersm all TEE test are passing. It is time to merge this, right? Please. 😢

wainersm commented:
Hi @BbolroC !

> @wainersm all TEE tests are passing. It is time to merge this, right? Please. 😢

I'd like to propose a change and it will trigger all jobs again....

... just kidding! Looks good! :)
